### PR TITLE
Speed up compilation of huge constant arrays

### PR DESCRIPTION
**Pull request by nnethercote**
@oli-obk mentioned a [toy program](https://github.com/rust-lang/rust/blob/c2f4744d2db4e162df824d0bd0b093ba4b351545/src/test/run-pass/mir_heavy_promoted.rs) recently on IRC, which contains this slow-to-compile constant:

> const TEST_DATA: [u8; 32 * 1024 * 1024] = [42; 32 * 1024 * 1024];

The attached commits roughly double compilation speed of this. The first commit does some inlining, which should be uncontroversial. The second commit introduces `write_values_to_ptr`, which more or less duplicates `write_value_to_ptr`, while moving a chunk of the code outside the hot loop. There is more room for improvement here by pushing the loop down some more, but I'm not sure how realistic this program is and thus whether it's worth the effort.

Also, the code duplication is unfortunate. One possibility would be to replace `write_value_to_ptr()` calls with `write_values_to_ptr(1)`, and remove `write_value_to_ptr()` altogether.

Thoughts?  CC @rust-lang/wg-compiler-performance 

(This PR probably shouldn't land in it's current form, so I'll mark it with r? @nnethercote so nobody else is pinged for review.)

### CODE REVIEW

**Comment by oli-obk** [comment]
While this is an improvement, the amount of code duplication is very unfortunate.

I think there are two avenues for exploration here:

1. Make repeat expressions fill in the first element and then have a method on `Memory` that fills in the remaining elements by copying the bits, undefinedness and relocations directly.

2. Optimize `Allocation`s that have many repeating parts by not storing the actual memory and instead doing some sort of sparse initialization as long as that takes less memory than the full initialization.

File: src/librustc_target/abi/mod.rs [hunk 1]
```diff
@@ -374,18 +380,22 @@ impl Align {
...
     pub fn abi(self) -> u64 {
         1 << self.abi_pow2
     }
 
+    #[inline]
     pub fn pref(self) -> u64 {
         1 << self.pref_pow2
     }
 
+    #[inline]
```
Comment (eddyb) [inline src/librustc_target/abi/mod.rs]: Oops!

**Comment by eddyb** [comment]
> Make repeat expressions fill in the first element and then have a method on `Memory` that fills in the remaining elements by copying the bits, undefinedness and relocations directly.

I like this! It's a lot like what I want `Rvalue::Repeat` to actually be in MIR, that is:
```rust
dest[0] = foo();
dest[1..] = dest[0];
```

**Comment by nnethercote** [comment]
I would still appreciate advice on the real-world relevance of this code. I don't want to spend time on this if it will only speed up pathological microbenchmarks.

**Comment by nnethercote** [comment]
> I was planning on doing this either way.

I'm happy to let you take this over, if that's what you're suggesting.

**Comment by oli-obk** [comment]
> I'm happy to let you take this over, if that's what you're suggesting.

I meant that I want this to happen. As you can see from the comment, I added it over a year ago. Expect me to need another year to get to it.

**Comment by wesleywiser** [comment]
@nnethercote I'd be happy to work on this if you want to hand it off to somebody else

**Comment by nnethercote** [comment]
> @nnethercote I'd be happy to work on this if you want to hand it off to somebody else

Please do! Feel free to take over this PR, close it in favour of a different one, or whatever makes sense. I think the existing patch demonstrates the basics, in particular which part of the code is hot, but I'm happy to answer questions if anything is unclear.

**Comment by nnethercote** [comment]
And if you're on Linux, I recommend using rustc-perf to profile. Instructions are at https://github.com/rust-lang-nursery/rustc-perf/tree/master/collector#profiling-local-builds. For the `profile` subcommand you can add a new benchmark simply by placing it alongside existing benchmarks under `rustc-perf/collector/benchmarks/`. (Though that won't work with the `bench_local` benchmark, alas.) You can use `--filter` to restrict profiling to a single benchmark.

**Comment by stokhos** [comment]
This PR is replaced by #51833.

**Pull request closed without merging**